client/core/bot.go

```diff
@@ -0,0 +1,1712 @@
+
+// syncOraclePrice retreives price data for the makerBot's market and
+// and returns the volume-weighted average price.
+func (m *makerBot) syncOraclePrice(ctx context.Context) (float64, error) {
+	_, price, err := m.Core.marketReport(ctx, m.base.SupportedAsset, m.quote.SupportedAsset)
+	return price, err
+}
+
+// cachedOraclePrice gets the cached price data for the specified market.
+// Expired data will not be returned.
+func (c *Core) cachedOraclePrice(mktName string) *stampedPrice {
+	c.mm.cache.RLock()
+	defer c.mm.cache.RUnlock()
+	p := c.mm.cache.prices[mktName]
+	if p == nil {
+		return nil
+	}
+	if time.Since(p.stamp) > oraclePriceExpiration {
+		return nil
+	}
+	return p
+}
+
+// startOracleSync starts the syncOracle loop.
+func (m *makerBot) startOracleSync(ctx context.Context) {
+	m.wg.Add(1)
+	go func() {
+		m.syncOracle(ctx)
+		m.wg.Done()
+	}()
+}
+
+// syncOracle keeps the oracle price data updated by querying the data on
+// a regular interval.
+// TODO: We should only run one syncOracle per market pair, not per bot or per
+// market.
+func (m *makerBot) syncOracle(ctx context.Context) {
+	if !atomic.CompareAndSwapUint32(&m.oracleRunning, 0, 1) {
+		m.log.Errorf("syncOracle called while already running")
+		return
+	}
+	defer atomic.StoreUint32(&m.oracleRunning, 0)
+
+	var lastCheck time.Time
+	m.mm.cache.RLock()
+	if p := m.mm.cache.prices[m.market.Name]; p != nil {
+		lastCheck = p.stamp
+	}
+	m.mm.cache.RUnlock()
+
+	// If we don't already have a price, sync now.
+	if time.Since(lastCheck) > time.Minute {
+		if _, err := m.syncOraclePrice(ctx); err != nil {
+			m.log.Errorf("failed to establish a market-averaged price: %v", err)
+		}
+	}
+
+	ticker := time.NewTicker(oracleRecheckInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+		case <-ctx.Done():
+			return
+		}
+		if _, err := m.syncOraclePrice(ctx); err != nil {
+			m.log.Errorf("failed to get market-averaged price: %v", err)
+		}
+	}
+}
+
+// basisPrice is the basis price for the makerBot's market.
+func (m *makerBot) basisPrice() uint64 {
+	pgm := m.program()
+	midGap, err := m.book.MidGap()
+	if err != nil && !errors.Is(err, orderbook.ErrEmptyOrderbook) {
+		m.log.Errorf("error calculating mid-gap: %w", err)
+		return 0
+	}
+	if p := basisPrice(pgm.Host, m.market, pgm.OracleBias, pgm.oracleWeighting(), midGap, m, m.log); p > 0 {
+		return p
+	}
+	return m.market.ConventionalRateToMsg(pgm.EmptyMarketRate)
+}
+
+type basisPricer interface {
+	cachedOraclePrice(mktName string) *stampedPrice
+	fiatConversions() map[uint32]float64
+}
+
+// basisPrice is the oracle-weighted price on which to base the target price and
+// spread. The oracle weighting and bias is applied, but no other adjustments.
+func basisPrice(host string, mkt *Market, oracleBias, oracleWeighting float64, midGap uint64, bp basisPricer, log dex.Logger) uint64 {
+	basisPrice := float64(midGap) // float64 message-rate units
+
+	log.Tracef("basisPrice: mid-gap price = %d", midGap)
+
```

Comment:
We should be a little more careful with using the current mid-gap here. If the book is sparse, someone could manipulate the bot into providing a favorable price by placing orders far from the gap. We should try to get some metric of order book health before using the mid-gap as an input, and we should probably only allow the mid-gap to vary so much from the oracle rate.